quickwit-metastore/src/lib.rs

```diff
@@ -59,3 +63,68 @@ pub use metastore::{
     IndexMetadata, MetadataSet, Metastore, SplitMetadata, SplitMetadataAndFooterOffsets, SplitState,
 };
 pub use metastore_resolver::{MetastoreFactory, MetastoreUriResolver};
+
+/// A trait to allow for our data model to check itself
+#[async_trait]
+pub trait Check {
+    /// An async function that implements the check on this type
+    async fn check(&self) -> Result<(), anyhow::Error>;
+}
+
+#[cfg(any(test, feature = "testsuite"))]
+#[async_trait]
+impl Check for MockMetastore {
+    async fn check(&self) -> anyhow::Result<()> {
+        Ok(())
+    }
+}
+
+/// Decorate a result as a check (for use in [`do_checks`]).
+///
+/// This takes out the original error of the `Result` (if any), replacing it with "taken".
+/// This measure is unfortunately required by the fact that `anyhow::Error` is not `Clone`.
+/// As the checks should bail anyway, this should not be a problem.
+pub fn checked<T>(r: &mut anyhow::Result<T>) -> Pin<Box<(dyn Future<Output = anyhow::Result<()>> + Send)>> {
+    let res = match r {
+        Ok(_) => Ok(()),
+        Err(e) => Err(std::mem::replace(e, anyhow::anyhow!("taken"))),
+    };
+    Box::pin(std::future::ready(res))
+}
+
+/// Decorate a check with nice output
+pub fn map_results<F: Future<Output = anyhow::Result<()>>>(
+    description: &str,
+    f: F,
+) -> impl Future<Output = Option<anyhow::Error>> {
+    let description = description.to_owned();
+    f.map(|result| match result {
+        Ok(()) => {
+            eprintln!("✅ {}", description);
+            None
+        }
+        Err(e) => {
+            eprintln!("❌ {}", description);
+            Some(e.context(description))
+        }
+    })
+}
+
+/// Run a list of early checks
+pub async fn do_checks<'a>(checks: impl IntoIterator<Item = (&'_ str, Pin<Box<(dyn Future<Output = anyhow::Result<()>> + Send + 'a)>>)>) -> anyhow::Result<()> {
```

Comment:
Generics are overkill too :) 
Vec or array would have been ok.
&'static str might have been ok too.

Reply:
I got a type error when trying to use `'static` here because it appears to get casted away by the async method lowering. I'll try using a Vec.